download_plugins.py

```diff
@@ -1,84 +1,124 @@
+            if dependency["name"] not in multi_version_plugins:
+                    multi_version_plugins.append(dependency["name"])
 
-            download_plugin(dependency['name'])
-            get_dependencies(dependency['name'])
         else:
-            if dependency['optional']:
-                print dependency['name'] + " is optional"
-            else: 
-                print dependency['name'] + " already downloaded"
+            print("Adding dependency: {}".format(dependency["name"]))
+            stored_plugins[dependency["name"]] = dependency["version"]
+            get_dependencies(dependency["name"])
 
-# Download plugin from jenkins update server. If no version is specified, latest plugin is downloaded.
-def download_plugin(plugin, version = None):
+
+# Download plugin from jenkins update server.
+# If no version is specified, latest plugin is downloaded.
+def download_plugin(plugin, version=None):
     download_url = ""
-    
+
     if version is None:
-        download_url = plugin_base_url + "/latest/%s.hpi" % plugin
+        download_url = (plugin_base_url +
+                        "/latest/{}.hpi".format(plugin))
     else:
-        download_url = plugin_base_url + "/download/plugins/%(plugin)s/%(version)s/%(plugin)s.hpi" % {'plugin': plugin, 'version': version}
+        download_url = (plugin_base_url +
+                        "/download/plugins/" +
+                        "{plugin}/{version}/{plugin}.hpi".format(
+                            plugin=plugin, version=version
+                        ))
 
     plugin_download = requests.get(download_url, stream=True)
 
-    if plugin_download.status_code is 200:
+    if plugin_download.status_code == 200:
         destination_path = download_directory + "/" + plugin + ".hpi"
 
-        with open(destination_path, 'wb') as data:
+        with open(destination_path, "wb") as data:
             for chunk in plugin_download.iter_content(chunk_size=128):
                 data.write(chunk)
 
-        print "Downloaded", plugin
-        
+        print("Downloaded", plugin)
+
     else:
-        print "Error downloading %s. Response:" % plugin
-        print plugin_download.text
+        print("Error downloading {}. Response:".format(plugin))
+        print(plugin_download.text)
+        global exit_code
         exit_code = 1
 
-    installed_plugins.append(plugin)
 
 # Install each plugin in the supplied json file along with dependencies.
 def install_plugins():
     for plugin, version in plugins.items():
-        print "**** Install Plugin: %s ****" % plugin
+        print("\n**** Install Plugin: {} *****".format(plugin))
         download_plugin(plugin, version)
         get_dependencies(plugin)
 
+    print("\nDownloading all dependencies")
+    print("*****************************************")
+    for plugin, version in stored_plugins.items():
+        download_plugin(plugin, version)
+
+    if multi_version_plugins:
```

Comment:
Nice job on having this logic to show the plugins which have multiple versions. It will be super helpful when we run into the version conflict issue in the future.
Just my two cents, is it possible to list all versions for these plugins? For example:

```
Warning: Multiple versions of [junit] found in dependencies.
    Plugin [A] requires [junit] version 1.19
    Plugin [B] requires [junit] version 1.20
    Version 1.20 downloaded
```